DataFormats/Common/interface/StdArray.h

```diff
@@ -0,0 +1,307 @@
+  public:
+    // Member types
+
+    using value_type = T;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using reference = value_type&;
+    using const_reference = value_type const&;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using iterator = pointer;
+    using const_iterator = const_pointer;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    // Constructors
+
+    // default constructor
+    constexpr StdArray() = default;
+
+    // copy constructor
+    constexpr StdArray(StdArray<T, N> const& init) = default;
+
+    // move constructor
+    constexpr StdArray(StdArray<T, N>&& init) = default;
+
+    // copy assignment operator
+    constexpr StdArray& operator=(StdArray<T, N> const& init) = default;
+
+    // move assignment operator
+    constexpr StdArray& operator=(StdArray<T, N>&& init) = default;
+
+    // Interoperability with std::array
+
+    /*
+    // constructor from an std::array
+    constexpr StdArray(std::array<T, N> const& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = init[i];
+      }
+    }
+
+    // constructor from an std::array temporary
+    constexpr StdArray(std::array<T, N>&& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = std::move(init[i]);
+      }
+    }
+    */
+
+    // copy assignment from an std::array
+    constexpr StdArray& operator=(std::array<T, N> const& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = init[i];
+      }
+      return *this;
+    }
+
+    // move assignment from an std::array
+    constexpr StdArray& operator=(std::array<T, N>&& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = std::move(init[i]);
+      }
+      return *this;
+    }
+
+    // cast operator to an std::array
+    constexpr operator std::array<T, N>() const {
+      std::array<T, N> copy;
+      for (size_type i = 0; i < N; ++i) {
+        copy[i] = data_[i];
+      }
+      return copy;
+    }
+
+    // Element access
+
+    // Returns a reference to the element at specified location pos, with bounds checking.
+    // If pos is not within the range of the container, an exception of type std::out_of_range is thrown.
+    constexpr reference at(size_type pos) {
+      if (pos >= N)
+        abort();
+      return data_[pos];
+    }
+    constexpr const_reference at(size_type pos) const {
+      if (pos >= N)
+        abort();
+      return data_[pos];
+    }
+
+    // Returns a reference to the element at specified location pos. No bounds checking is performed.
+    constexpr reference operator[](size_type pos) { return data_[pos]; }
+    constexpr const_reference operator[](size_type pos) const { return data_[pos]; }
+
+    // Returns a reference to the first element in the container.
+    // Calling front on an empty container causes undefined behavior.
```

Comment:
Yes, I'd prefer to catch such a but sooner rather than later.
Also, what should we return otherwise ?
I'll update the comment.

Reply:
> I'll update the comment.

Ok.

> Also, what should we return otherwise ?

**If** the contract would be "undefined behavior" (which it isn't anymore), returning e.g. `data_[0]` would be according to the contract.

Reply:
For the empty case, `data_` is an empty `struct`, so `data_[0]` does not even compile.

OK, it could have been implemented differently, with a 1 element array.

Or we could return `T{}`.
But I prefer to fail explicitly.